marlowe-benchmark/app/Language/Marlowe/Runtime/Benchmark/Query.hs

```diff
@@ -0,0 +1,113 @@
+{-# LANGUAGE RecordWildCards #-}
+
+-- | Benchmark for the `Query` protocol.
+module Language.Marlowe.Runtime.Benchmark.Query (
+  -- * Benchmarking
+  Benchmark (..),
+  measure,
+) where
+
+import Control.Monad.IO.Class (MonadIO, liftIO)
+import Control.Monad.Trans.Marlowe (MarloweT)
+import Data.Aeson (ToJSON)
+import Data.Default (Default (..))
+import Data.Foldable (foldlM)
+import Data.List.Split (chunksOf)
+import Data.Time.Clock (NominalDiffTime)
+import Data.Time.Clock.POSIX (getPOSIXTime)
+import GHC.Generics (Generic)
+import Language.Marlowe.Protocol.Query.Client (MarloweQueryClient, getContractHeaders)
+import Language.Marlowe.Protocol.Query.Types (ContractFilter, Order (Ascending), Page (..), Range (..))
+import Language.Marlowe.Runtime.Client (runMarloweQueryClient)
+import UnliftIO (forConcurrently)
+
+data Benchmark = Benchmark
+  { metric :: String
+  , query :: String
+  , queriesPerSecond :: Double
+  , pagesPerSecond :: Double
+  , contractsPerSecond :: Double
+  , seconds :: Double
+  }
+  deriving (Eq, Generic, Ord, Show, ToJSON)
+
+data Statistics = Statistics
+  { queries :: Int
+  , pages :: Int
+  , contracts :: Int
+  , duration :: NominalDiffTime
+  }
+  deriving (Eq, Generic, Ord, Show, ToJSON)
+
+instance Default Statistics where
+  def = Statistics def def def 0
+
+-- | Measure the performance of the protocol.
+measure
+  :: Int
+  -- ^ Number of parallel clients for `Query` protocol.
+  -> Int
+  -- ^ Number of queries to be executed by each `Query` client.
+  -> Int
+  -- ^ Page size for each `Query` client.
+  -> String
+  -- ^ Label for the contract filter.
+  -> [ContractFilter]
+  -- ^ Contract filters to query.
+  -> MarloweT IO [Benchmark]
+  -- ^ Action to run the benchmark.
+measure parallelism batchSize pageSize query filters =
+  let batches = take parallelism $ chunksOf batchSize filters
+   in forConcurrently batches $
+        run "Query" pageSize query
+
+-- | Run a benchmark client.
+run
+  :: String
+  -- ^ Label for the benchmark.
+  -> Int
+  -- ^ Page size for each `Query` client.
+  -> String
+  -- ^ Label for the contract filter.
+  -> [ContractFilter]
+  -- ^ Contract filters to query.
+  -> MarloweT IO Benchmark
+  -- ^ Action to run the benchmark.
+run metric pageSize query filters =
+  do
+    start <- liftIO getPOSIXTime
+    Statistics{..} <- foldlM ((runMarloweQueryClient .) . benchmark start pageSize) def filters
```

Comment:
Note because `MarloweQueryClient` is a monad this could be

```suggestion
    Statistics{..} <- runMarloweQueryClient $ foldlM (benchmark start pageSize) def filters
```

The difference is that you don't establish a new connection for each filter. I'll leave it up to you to decide if this is what you want or not.

Reply:
I kept that `foldM` outermost because I thought that might be a bit more stressful to the backend service. Normally, I would have put `runMarloweQueryClient` on the outside, for efficiency.